File: Kth Smallest Element in a Sorted Matrix.cpp
class Solution {
public:
    int kthSmallest(vector<vector<int>>& matrix, int k) {
        int n = matrix.size();
        int a,b;
        int m = n*n;
        vector<int> linear;
        for(int i = 0 ; i<m ; i++)
        {
            a = i/n;
            b = i%n;
            linear.push_back(matrix[a][b]);
        }
        make_heap(linear.begin(),linear.end(),greater());
        for(int i = 1 ; i<k ; i ++)
        {
            pop_heap(linear.begin(),linear.end(),greater());
            linear.pop_back();
        }
        return linear[0];
    }
};

File: kth-smallest-element-in-a-sorted-matrix.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:

        def leftChild(i):
            return 2 * i + 1


        def rightChild(i):
            return 2 * i + 2

        def heapify_down(arr, n, current):
            left = leftChild(current)
            right = rightChild(current)

            if left < n and arr[left] < arr[current]:
                arr[current], arr[left] = arr[left], arr[current]
                heapify_down(arr, n, left)

            if right < n and arr[right] < arr[current]:
                arr[current], arr[right] = arr[right], arr[current]
                heapify_down(arr, n, right)

        heapify(matrix)

        for _ in range(k-1):
            heappop(matrix[0])
            heapify_down(matrix,len(matrix),0)
            if not len(matrix[0]):
                heappop(matrix)

        return matrix[0][0]